### ISSUE TITLE
False positive `attribute-defined-outside-init` error in Python 3.6 when attribute is defined in generic base class

### ISSUE THREAD

**Issue by huzecong** [pylint-dev/pylint#2981]
### Steps to reproduce
1. Create file `foo.py` containing:

   ```python
   from typing import Generic, TypeVar
   
   T = TypeVar('T')
   
   class Base(Generic[T]):
       def __init__(self):
           self.val = False
   
   class Derived(Base[T]):
       def func(self):
           self.val = True
   
   a = Derived()
   a.func()
   print(a.val)
   ```

2. Run `pylint foo.py` in a Python 3.6 envinroment.

### Current behavior
Pylint reports:
```
foo.py:11:8: W0201: Attribute 'val' defined outside __init__ (attribute-defined-outside-init)
```

It is also noted that the error is **not** reported when:

1. Using Python versions 3.7+.
2. Removing the `Generic[T]` base class in `Base`.
3. Changing base class of `Derived` from `Base[T]` to `Base`.

### Expected behavior
No errors should be reported because `val` is defined in `__init__` method of base class.

### pylint --version output
```
pylint 2.3.1
astroid 2.2.5
Python 3.6.7 (default, May 29 2019, 10:02:40)
[GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.46.4)]
```

**Comment by PCManticore** [comment]
Thanks for the report.

**Comment by tucked** [comment]
We can force a false `access-member-before-definition` pretty easily here too:
``` diff
diff --git a/foo.py b/bar.py
index d0b7735..1225dc0 100644
--- a/foo.py
+++ b/bar.py
@@ -8,6 +8,7 @@ class Base(Generic[T]):

 class Derived(Base[T]):
     def func(self):
+        print(self.val)
         self.val = True

 a = Derived()
```
```
bar.py:11:14: E0203: Access to member 'val' before its definition line 12 (access-member-before-definition)
```
Interestingly, this does NOT repro:
``` diff
diff --git a/foo.py b/bar.py
index d0b7735..8d17d0f 100644
--- a/foo.py
+++ b/bar.py
@@ -8,7 +8,7 @@ class Base(Generic[T]):

 class Derived(Base[T]):
     def func(self):
-        self.val = True
+        print(self.val)

 a = Derived()
 a.func()
```

**Comment by mthuurne** [comment]
With the following test case, I can reproduce the problem on Python 3.8 as well:
```py
from typing import Generic, TypeVar

T = TypeVar('T')

class Base(Generic[T]):
    def __init__(self, x: T):
        self.x = x

class Middle(Base[T]):
    pass

class Derived(Middle[int]):
    def set_x(self, x: int) -> None:
        self.x = x
```
Even though the `Middle` class contains no functionality, it is essential for reproducing: if `Derived` inherits from `Base` directly, the `attribute-defined-outside-init` message is not reported (when using Python 3.8).

Exact versions used:
```
pylint 2.4.4
astroid 2.3.3
Python 3.8.2 (default, Mar 05 2020, 18:58:42) [GCC]
```

**Comment by Pierre-Sassoulas** [comment]
Closing as it was fixed at some point before 2.14.

**Comment by HWiese1980** [comment]
I see the regression too. Python 3.9.10, PyLint 3.0.0a7. Exact same behavior as described above. 

What's the status here?

**Comment by Pierre-Sassoulas** [comment]
It needs a PR, any contributions are welcome :)

**Comment by Pierre-Sassoulas** [comment]
I cannot reproduce this, and python 3.10 is EOL, so if you encounter something similar please reopen with a current reproducer.
